From 133f714951148c680f92f54eef8fcb31d3b76921 Mon Sep 17 00:00:00 2001
From: Dmitry Ivanov <idmitry@qti.qualcomm.com>
Date: Thu, 15 Apr 2021 15:49:37 +0200
Subject: [PATCH] shaders: Modify ucompare statement template to work around a
 Qualcomm shader compiler bug.

The actual problem was the usage of certain functions on the guest
side in a virtio-gpu setup, where the shader code contained the step()
function. After translation to the internal TGSI shader format, it
will be unpacked by virglrenderer into GLSL again, and the template in
use created a construct where the Qualcomm shader compiler had a
problem with.

The template originally creates constructs like this:

  float val = float(uintBitsToFloat(uint(greaterThanEqual(tmpA, vec4(0.8)).x) * uint(0xffffffff)));

which the new template generates as:

  float val = float(greaterThanEqual(tmpA, vec4(0.8)).x);
  val = float(uintBitsToFloat(uint(val) * uint(0xffffffff)));

This is (almost) the same, but the former doesn't work on Qualcomm
systems, while the latter is running without problems.

NOTE: Because it just reformulates the same thing, it doesn't rely on
      Qualcomm specific parts to be present, so can be used on
      non-Qualcomm platforms as well, even if that fix is not strictly
      necessary.

Signed-off-by: Dmitry Ivanov <idmitry@qti.qualcomm.com>
---
 src/vrend_shader.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/src/vrend_shader.c b/src/vrend_shader.c
index bf3873fa..90108570 100644
--- a/src/vrend_shader.c
+++ b/src/vrend_shader.c
@@ -2520,7 +2520,23 @@ static void emit_fog_fixup_write(const struct dump_ctx *ctx,
 #define emit_op1(op) emit_buff(&ctx->glsl_strbufs, "%s = %s(%s(%s(%s))%s);\n", dsts[0], get_string(dinfo.dstconv), get_string(dinfo.dtypeprefix), op, srcs[0], writemask)
 #define emit_compare(op) emit_buff(&ctx->glsl_strbufs, "%s = %s(%s((%s(%s(%s), %s(%s))))%s);\n", dsts[0], get_string(dinfo.dstconv), get_string(dinfo.dtypeprefix), op, get_string(sinfo.svec4), srcs[0], get_string(sinfo.svec4), srcs[1], writemask)
 
-#define emit_ucompare(op) emit_buff(&ctx->glsl_strbufs, "%s = %s(uintBitsToFloat(%s(%s(%s(%s), %s(%s))%s) * %s(0xffffffff)));\n", dsts[0], get_string(dinfo.dstconv), get_string(dinfo.udstconv), op, get_string(sinfo.svec4), srcs[0], get_string(sinfo.svec4), srcs[1], writemask, get_string(dinfo.udstconv))
+#define emit_ucompare(op)                                       \
+    emit_buff(&ctx->glsl_strbufs,                               \
+        "%s = %s(%s(%s(%s), %s(%s))%s);\n"                      \
+        "%s = %s(uintBitsToFloat(%s(%s) * %s(0xffffffff)));\n", \
+        dsts[0],                                                \
+        get_string(dinfo.dstconv),                              \
+        op,                                                     \
+        get_string(sinfo.svec4),                                \
+        srcs[0],                                                \
+        get_string(sinfo.svec4),                                \
+        srcs[1],                                                \
+        writemask,                                              \
+        dsts[0],                                                \
+        get_string(dinfo.dstconv),                              \
+        get_string(dinfo.udstconv),                             \
+        dsts[0],                                                \
+        get_string(dinfo.udstconv))
 
 static void handle_vertex_proc_exit(const struct dump_ctx *ctx,
                                     struct vrend_glsl_strbufs *glsl_strbufs,
-- 
2.31.1

